### PR TITLE
[FLINK-16072][python] Optimize the performance of the write/read null mask

### PR DESCRIPTION
**Pull request by HuangXingBo**
## What is the purpose of the change

*This pr optimize the performance of the write/read null mask*

## Brief change log

  - *The decode/encode/write_null_mask/read_mask in FlattenRowCoder*

## Verifying this change

This change added tests and can be verified as follows:

  - *It is a perforce improvement without extra function test*

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (no)
  - The serializers: (no)
  - The runtime per-record code paths (performance sensitive): (no)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Yarn/Mesos, ZooKeeper: (no)
  - The S3 file system connector: (no)

## Documentation

  - Does this pull request introduce a new feature? (no)
  - If yes, how is the feature documented? (not applicable)

### Test Code

    @udf(input_types=[DataTypes.INT(False)], result_type=DataTypes.INT(False))
    def inc(x):
        return x

    t_env.register_function("inc", inc)

    # num_rows = 100000000
    num_rows = 100000
    num_columns = 100

    select_list = ["inc(c%s)" % i for i in range(num_columns)]
    t_env.register_table_sink(
        "sink",
        PrintTableSink(
            ["c%s" % i for i in range(num_columns)],
            [DataTypes.INT(False)] * num_columns))

    t_env.from_table_source(MultiRowColumnTableSource(num_rows, num_columns)) \
        .select(','.join(select_list)) \
        .insert_into("sink")

    beg_time = time.time()
    t_env.execute("perf_test")
    print("consume time: " + str(time.time() - beg_time))

## Test Results
num rows, num colums |  Consume Time(Before) | Consume Time(After)
1000w,10                       |     262.01                          |   222.07
100w, 100                      |     161.12                           |   148.23

### CODE REVIEW

File: flink-python/pyflink/fn_execution/coder_impl.py [hunk 1]
```diff
@@ -25,56 +25,331 @@
...
+                              (True, True, True, True, True, True, False, True),
+                              (True, True, True, True, True, True, True, False),
+                              (True, True, True, True, True, True, True, True))
+
+    null_byte_search_table = (0x80, 0x40, 0x20, 0x10, 0x08, 0x04, 0x02, 0x01)
 
     def __init__(self, field_coders):
         self._field_coders = field_coders
         self._filed_count = len(field_coders)
+        self._complete_byte_num = self._filed_count // 8
```
Comment (dianfu) [inline flink-python/pyflink/fn_execution/coder_impl.py]: What about renamed to _leading_bytes_num

File: flink-python/pyflink/fn_execution/coder_impl.py [hunk 2]
```diff
@@ -25,56 +25,331 @@
...
+        self._remaining_bit_num = self._filed_count % 8
 
     def encode_to_stream(self, value, out_stream, nested):
-        self.write_null_mask(value, out_stream)
+        self.write_null_mask(value, self._complete_byte_num, self._remaining_bit_num, out_stream)
```
Comment (dianfu) [inline flink-python/pyflink/fn_execution/coder_impl.py]: Make write_null_mask non-static? then this change is not needed any more.

File: flink-python/pyflink/fn_execution/coder_impl.py [hunk 3]
```diff
@@ -25,56 +25,331 @@
...
         for i in range(self._filed_count):
-            if value[i] is not None:
-                self._field_coders[i].encode_to_stream(value[i], out_stream, nested)
+            item = value[i]
+            if item is not None:
+                self._field_coders[i].encode_to_stream(item, out_stream, nested)
 
     def decode_from_stream(self, in_stream, nested):
-        null_mask = self.read_null_mask(self._filed_count, in_stream)
+        null_mask = self.read_null_mask(self._complete_byte_num, self._remaining_bit_num, in_stream)
```
Comment (dianfu) [inline flink-python/pyflink/fn_execution/coder_impl.py]: ditto

File: flink-python/pyflink/fn_execution/coder_impl.py [hunk 4]
```diff
@@ -25,56 +25,331 @@
 
 
 class FlattenRowCoderImpl(StreamCoderImpl):
+    null_mask_search_table = ((False, False, False, False, False, False, False, False),
```
[the remark this answers is not in the document]
Comment (HuangXingBo) [reply flink-python/pyflink/fn_execution/coder_impl.py]: Yes. Good idea

**Comment by HuangXingBo** [comment]
Thanks a lot for @dianfu review, I have addressed the comments at the latest commit.

File: flink-python/pyflink/fn_execution/coder_impl.py [hunk 5]
```diff
@@ -24,57 +24,71 @@
...
+    for b in range(num):
+        every_num_null_mask = [(b & 0x80) > 0, (b & 0x40) > 0, (b & 0x20) > 0, (b & 0x10) > 0,
+                               (b & 0x08) > 0, (b & 0x04) > 0, (b & 0x02) > 0, (b & 0x01) > 0]
+        null_mask.append(tuple(every_num_null_mask))
+
+    return tuple(null_mask)
+
+
 class FlattenRowCoderImpl(StreamCoderImpl):
+    null_mask_search_table = generate_null_mask_search_table()
```
Comment (dianfu) [inline flink-python/pyflink/fn_execution/coder_impl.py]: What about make it an instance variable as it's faster than static variable: https://stackoverflow.com/questions/2714573/instance-variables-vs-class-variables-in-python

File: flink-python/pyflink/fn_execution/coder_impl.py [hunk 6]
```diff
@@ -24,57 +24,71 @@
...
+                self._field_coders[i].encode_to_stream(item, out_stream, nested)
 
     def decode_from_stream(self, in_stream, nested):
-        null_mask = self.read_null_mask(self._filed_count, in_stream)
+        null_mask = self.read_null_mask(in_stream)
         return [None if null_mask[idx] else self._field_coders[idx].decode_from_stream(
             in_stream, nested) for idx in range(0, self._filed_count)]
 
-    @staticmethod
-    def write_null_mask(value, out_stream):
```
Comment (dianfu) [inline flink-python/pyflink/fn_execution/coder_impl.py]: Could we add some unit tests for the null mask related logic?

File: flink-python/pyflink/fn_execution/coder_impl.py (hunk 1 above)
Comment (dianfu) [reply flink-python/pyflink/fn_execution/coder_impl.py]: _complete_byte_num -> _leading_complete_bytes_num
_leading_bytes_num -> _remaining_bits_num

**Comment by dianfu** [comment]
@HuangXingBo Thanks for the update. LGTM.
